File: riveredge-frontend/src/apps/kuaizhizao/services/quality-qms.ts
import { apiRequest } from '../../../services/api';

export type QmsEvidenceLink = {
  ref_type: string;
  ref_id?: number | null;
  ref_code?: string | null;
  ref_name?: string | null;
  path?: string | null;
  note?: string | null;
};

export interface QmsSystemDocument {
  id: number;
  uuid?: string;
  document_code: string;
  title: string;
  doc_type: string;
  version: string;
  status: string;
  iso_clause?: string;
  iso_clause_id?: number | null;
  content?: string;
  file_url?: string;
  effective_at?: string;
  obsolete_at?: string;
  next_review_at?: string;
  owner_name?: string;
  evidence_links?: QmsEvidenceLink[];
  training_refs?: QmsEvidenceLink[];
  attachments?: any;
  remarks?: string;
  created_at?: string;
  updated_at?: string;
}

export interface QmsInternalAudit {
  id: number;
  uuid?: string;
  audit_code: string;
  title: string;
  audit_scope?: string;
  iso_clause?: string;
  iso_clause_id?: number | null;
  status: string;
  planned_date?: string;
  completed_date?: string;
  lead_auditor?: string;
  audit_team?: string;
  checklist?: string;
  findings?: string;
  conclusion?: string;
  finding_links?: QmsEvidenceLink[];
  training_refs?: QmsEvidenceLink[];
  calibration_refs?: QmsEvidenceLink[];
  attachments?: any;
  remarks?: string;
  created_at?: string;
  updated_at?: string;
}

export interface QmsManagementReview {
  id: number;
  uuid?: string;
  review_code: string;
  title: string;
  status: string;
  review_date?: string;
  chairperson?: string;
  attendees?: string;
  inputs_summary?: string;
  outputs_summary?: string;
  input_links?: QmsEvidenceLink[];
  training_refs?: QmsEvidenceLink[];
  calibration_refs?: QmsEvidenceLink[];
  attachments?: any;
  remarks?: string;
  created_at?: string;
  updated_at?: string;
}

export interface QmsIsoClause {
  id: number;
  uuid?: string;
  standard_code: string;
  clause_code: string;
  title: string;
  description?: string;
  parent_id?: number | null;
  sort_order?: number;
  is_active?: boolean;
  created_by_name?: string;
  updated_by_name?: string;
  created_at?: string;
  updated_at?: string;
}

export interface QmsIsoClauseTreeNode extends QmsIsoClause {
  children?: QmsIsoClauseTreeNode[];
}

export interface QmsIsoClauseComplianceSummary {
  effective_document_count: number;
  review_due_count: number;
  internal_audit_count: number;
  last_audit_date?: string | null;
  has_gap: boolean;
  no_effective_document: boolean;
  no_completed_audit: boolean;
  compliance_status: 'covered' | 'review_due' | 'gap';
}

type ListEnvelope<T> = { items: T[]; total: number };

export const qualityQmsApi = {
  systemDocuments: {
    list: (params?: Record<string, unknown>) =>
      apiRequest<ListEnvelope<QmsSystemDocument>>('/apps/kuaizhizao/qms/system-documents', {
        method: 'GET',
        params,
      }),
    get: (id: number) =>
      apiRequest<QmsSystemDocument>(`/apps/kuaizhizao/qms/system-documents/${id}`, { method: 'GET' }),
    create: (data: Partial<QmsSystemDocument>) =>
      apiRequest<QmsSystemDocument>('/apps/kuaizhizao/qms/system-documents', { method: 'POST', data }),
    update: (id: number, data: Partial<QmsSystemDocument>) =>
      apiRequest<QmsSystemDocument>(`/apps/kuaizhizao/qms/system-documents/${id}`, {
        method: 'PUT',
        data,
      }),
    publish: (id: number) =>
      apiRequest<QmsSystemDocument>(`/apps/kuaizhizao/qms/system-documents/${id}/publish`, {
        method: 'POST',
      }),
    obsolete: (id: number) =>
      apiRequest<QmsSystemDocument>(`/apps/kuaizhizao/qms/system-documents/${id}/obsolete`, {
        method: 'POST',
      }),
    delete: (id: number) =>
      apiRequest(`/apps/kuaizhizao/qms/system-documents/${id}`, { method: 'DELETE' }),
    reviewDueSummary: () =>
      apiRequest<{ due_count: number }>('/apps/kuaizhizao/qms/system-documents/review-due-summary', {
        method: 'GET',
      }),
  },
  internalAudits: {
    list: (params?: Record<string, unknown>) =>
      apiRequest<ListEnvelope<QmsInternalAudit>>('/apps/kuaizhizao/qms/internal-audits', {
        method: 'GET',
        params,
      }),
    get: (id: number) =>
      apiRequest<QmsInternalAudit>(`/apps/kuaizhizao/qms/internal-audits/${id}`, { method: 'GET' }),
    create: (data: Partial<QmsInternalAudit>) =>
      apiRequest<QmsInternalAudit>('/apps/kuaizhizao/qms/internal-audits', { method: 'POST', data }),
    update: (id: number, data: Partial<QmsInternalAudit>) =>
      apiRequest<QmsInternalAudit>(`/apps/kuaizhizao/qms/internal-audits/${id}`, {
        method: 'PUT',
        data,
      }),
    delete: (id: number) =>
      apiRequest(`/apps/kuaizhizao/qms/internal-audits/${id}`, { method: 'DELETE' }),
  },
  managementReviews: {
    list: (params?: Record<string, unknown>) =>
      apiRequest<ListEnvelope<QmsManagementReview>>('/apps/kuaizhizao/qms/management-reviews', {
        method: 'GET',
        params,
      }),
    get: (id: number) =>
      apiRequest<QmsManagementReview>(`/apps/kuaizhizao/qms/management-reviews/${id}`, {
        method: 'GET',
      }),
    create: (data: Partial<QmsManagementReview>) =>
      apiRequest<QmsManagementReview>('/apps/kuaizhizao/qms/management-reviews', {
        method: 'POST',
        data,
      }),
    update: (id: number, data: Partial<QmsManagementReview>) =>
      apiRequest<QmsManagementReview>(`/apps/kuaizhizao/qms/management-reviews/${id}`, {
        method: 'PUT',
        data,
      }),
    delete: (id: number) =>
      apiRequest(`/apps/kuaizhizao/qms/management-reviews/${id}`, { method: 'DELETE' }),
    inputSummary: (params?: { period_start?: string; period_end?: string }) =>
      apiRequest<{
        summary_text: string;
        nonconforming_count: number;
        open_8d_count: number;
        iqc_pass_rate?: number | null;
        oqc_pass_rate?: number | null;
      }>('/apps/kuaizhizao/qms/management-reviews/input-summary', { method: 'GET', params }),
  },
  isoClauses: {
    list: (params?: Record<string, unknown>) =>
      apiRequest<ListEnvelope<QmsIsoClause>>('/apps/kuaizhizao/qms/iso-clauses', {
        method: 'GET',
        params,
      }),
    tree: (params?: { standard_code?: string }) =>
      apiRequest<QmsIsoClauseTreeNode[]>('/apps/kuaizhizao/qms/iso-clauses/tree', {
        method: 'GET',
        params,
      }),
    get: (id: number) =>
      apiRequest<QmsIsoClause>(`/apps/kuaizhizao/qms/iso-clauses/${id}`, { method: 'GET' }),
    create: (data: Partial<QmsIsoClause>) =>
      apiRequest<QmsIsoClause>('/apps/kuaizhizao/qms/iso-clauses', { method: 'POST', data }),
    update: (id: number, data: Partial<QmsIsoClause>) =>
      apiRequest<QmsIsoClause>(`/apps/kuaizhizao/qms/iso-clauses/${id}`, {
        method: 'PUT',
        data,
      }),
    delete: (id: number) =>
      apiRequest(`/apps/kuaizhizao/qms/iso-clauses/${id}`, { method: 'DELETE' }),
    loadPreset: (standard_code = 'ISO9001:2015') =>
      apiRequest<{ created: number; skipped: number; linked: number }>(
        '/apps/kuaizhizao/qms/iso-clauses/load-preset',
        { method: 'POST', params: { standard_code } },
      ),
    complianceSummary: (id: number) =>
      apiRequest<QmsIsoClauseComplianceSummary>(
        `/apps/kuaizhizao/qms/iso-clauses/${id}/compliance-summary`,
        { method: 'GET' },
      ),
    relatedDocuments: (id: number) =>
      apiRequest<{ items: QmsSystemDocument[] }>(
        `/apps/kuaizhizao/qms/iso-clauses/${id}/related-documents`,
        { method: 'GET' },
      ),
    relatedAudits: (id: number) =>
      apiRequest<{ items: QmsInternalAudit[] }>(
        `/apps/kuaizhizao/qms/iso-clauses/${id}/related-audits`,
        { method: 'GET' },
      ),
  },
};

File: riveredge-backend/src/apps/kuaizhizao/schemas/quality_qms.py
"""质量体系 schemas：体系文件 / 内审 / 管理评审"""

from __future__ import annotations

from datetime import datetime
from typing import Any, List, Optional

from pydantic import Field

from core.schemas.base import BaseSchema


class QmsEvidenceLink(BaseSchema):
    ref_type: str = Field(..., description="证据类型")
    ref_id: Optional[int] = None
    ref_code: Optional[str] = None
    ref_name: Optional[str] = None
    path: Optional[str] = None
    note: Optional[str] = None


class QmsSystemDocumentBase(BaseSchema):
    document_code: Optional[str] = Field(None, description="文件编码")
    title: str = Field(..., description="标题")
    doc_type: str = Field("procedure", description="文件类型")
    version: str = Field("A0", description="版本")
    status: str = Field("draft", description="状态")
    iso_clause: Optional[str] = None
    iso_clause_id: Optional[int] = None
    content: Optional[str] = None
    file_url: Optional[str] = None
    effective_at: Optional[datetime] = None
    obsolete_at: Optional[datetime] = None
    next_review_at: Optional[datetime] = None
    owner_name: Optional[str] = None
    evidence_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None


class QmsSystemDocumentCreate(QmsSystemDocumentBase):
    pass


class QmsSystemDocumentUpdate(BaseSchema):
    title: Optional[str] = None
    doc_type: Optional[str] = None
    version: Optional[str] = None
    status: Optional[str] = None
    iso_clause: Optional[str] = None
    iso_clause_id: Optional[int] = None
    content: Optional[str] = None
    file_url: Optional[str] = None
    effective_at: Optional[datetime] = None
    obsolete_at: Optional[datetime] = None
    next_review_at: Optional[datetime] = None
    owner_name: Optional[str] = None
    evidence_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None
    document_code: Optional[str] = None


class QmsSystemDocumentResponse(QmsSystemDocumentBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class QmsSystemDocumentListResponse(BaseSchema):
    items: List[QmsSystemDocumentResponse]
    total: int


class QmsInternalAuditBase(BaseSchema):
    audit_code: Optional[str] = None
    title: str
    audit_scope: Optional[str] = None
    iso_clause: Optional[str] = None
    iso_clause_id: Optional[int] = None
    status: str = Field("planned")
    planned_date: Optional[datetime] = None
    completed_date: Optional[datetime] = None
    lead_auditor: Optional[str] = None
    audit_team: Optional[str] = None
    checklist: Optional[str] = None
    findings: Optional[str] = None
    conclusion: Optional[str] = None
    finding_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    calibration_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None


class QmsInternalAuditCreate(QmsInternalAuditBase):
    pass


class QmsInternalAuditUpdate(BaseSchema):
    title: Optional[str] = None
    audit_scope: Optional[str] = None
    iso_clause: Optional[str] = None
    iso_clause_id: Optional[int] = None
    status: Optional[str] = None
    planned_date: Optional[datetime] = None
    completed_date: Optional[datetime] = None
    lead_auditor: Optional[str] = None
    audit_team: Optional[str] = None
    checklist: Optional[str] = None
    findings: Optional[str] = None
    conclusion: Optional[str] = None
    finding_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    calibration_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None
    audit_code: Optional[str] = None


class QmsInternalAuditResponse(QmsInternalAuditBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class QmsInternalAuditListResponse(BaseSchema):
    items: List[QmsInternalAuditResponse]
    total: int


class QmsManagementReviewBase(BaseSchema):
    review_code: Optional[str] = None
    title: str
    status: str = Field("draft")
    review_date: Optional[datetime] = None
    chairperson: Optional[str] = None
    attendees: Optional[str] = None
    inputs_summary: Optional[str] = None
    outputs_summary: Optional[str] = None
    input_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    calibration_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None


class QmsManagementReviewCreate(QmsManagementReviewBase):
    pass


class QmsManagementReviewUpdate(BaseSchema):
    title: Optional[str] = None
    status: Optional[str] = None
    review_date: Optional[datetime] = None
    chairperson: Optional[str] = None
    attendees: Optional[str] = None
    inputs_summary: Optional[str] = None
    outputs_summary: Optional[str] = None
    input_links: Optional[List[Any]] = None
    training_refs: Optional[List[Any]] = None
    calibration_refs: Optional[List[Any]] = None
    attachments: Optional[Any] = None
    remarks: Optional[str] = None
    review_code: Optional[str] = None


class QmsManagementReviewResponse(QmsManagementReviewBase):
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class QmsManagementReviewListResponse(BaseSchema):
    items: List[QmsManagementReviewResponse]
    total: int


class QmsSystemDocumentReviewDueSummary(BaseSchema):
    due_count: int = Field(..., description="待复审文件数（生效中且 next_review_at 已到期或为空不计）")


class QmsManagementReviewInputSummary(BaseSchema):
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    nonconforming_count: int = 0
    open_8d_count: int = 0
    iqc_total: int = 0
    iqc_pass_count: int = 0
    iqc_pass_rate: Optional[float] = None
    oqc_total: int = 0
    oqc_pass_count: int = 0
    oqc_pass_rate: Optional[float] = None
    summary_text: str = ""


class QmsIsoClauseBase(BaseSchema):
    standard_code: str = Field(..., description="标准编码")
    clause_code: str = Field(..., description="条款号")
    title: str = Field(..., description="条款标题")
    description: Optional[str] = None
    parent_id: Optional[int] = None
    sort_order: int = Field(0, description="排序")
    is_active: bool = Field(True, description="是否启用")


class QmsIsoClauseCreate(QmsIsoClauseBase):
    pass


class QmsIsoClauseUpdate(BaseSchema):
    standard_code: Optional[str] = None
    clause_code: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    parent_id: Optional[int] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class QmsIsoClauseResponse(QmsIsoClauseBase):
    id: int
    uuid: str
    tenant_id: int
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class QmsIsoClauseListResponse(BaseSchema):
    items: List[QmsIsoClauseResponse]
    total: int


class QmsIsoClauseTreeNode(QmsIsoClauseResponse):
    children: List["QmsIsoClauseTreeNode"] = Field(default_factory=list)


class QmsIsoClauseComplianceSummary(BaseSchema):
    effective_document_count: int = 0
    review_due_count: int = 0
    internal_audit_count: int = 0
    last_audit_date: Optional[datetime] = None
    has_gap: bool = False
    no_effective_document: bool = False
    no_completed_audit: bool = False
    compliance_status: str = Field("gap", description="covered|review_due|gap")


class QmsIsoClauseLoadPresetResult(BaseSchema):
    created: int = 0
    skipped: int = 0
    linked: int = 0


class QmsIsoClauseRelatedDocumentsResponse(BaseSchema):
    items: List[QmsSystemDocumentResponse]


class QmsIsoClauseRelatedAuditsResponse(BaseSchema):
    items: List[QmsInternalAuditResponse]


QmsIsoClauseTreeNode.model_rebuild()
